In [1]:
import pandas as pd 
import numpy as np

In [2]:
df = pd.read_csv('../data/hotel_bookings.csv')

In [3]:
data = df.copy()

In [4]:
data.shape

(119390, 32)

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [6]:
data.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [7]:
data.duplicated().sum()

np.int64(31994)

### Working on Duplicates

In [8]:
data[data.duplicated()].head(10)

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
5,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.00,0,1,Check-Out,2015-07-03
22,Resort Hotel,0,72,2015,July,27,1,2,4,2,...,No Deposit,250.0,NaN,0,Transient,84.67,0,1,Check-Out,2015-07-07
43,Resort Hotel,0,70,2015,July,27,2,2,3,2,...,No Deposit,250.0,NaN,0,Transient,137.00,0,1,Check-Out,2015-07-07
138,Resort Hotel,1,5,2015,July,28,5,1,0,2,...,No Deposit,240.0,NaN,0,Transient,97.00,0,0,Canceled,2015-07-01
200,Resort Hotel,0,0,2015,July,28,7,0,1,1,...,No Deposit,240.0,NaN,0,Transient,109.80,0,3,Check-Out,2015-07-08
219,Resort Hotel,1,1,2015,July,28,8,0,1,2,...,No Deposit,NaN,110.0,0,Transient,104.72,0,1,Canceled,2015-07-08
256,Resort Hotel,0,91,2015,July,28,10,0,2,2,...,No Deposit,240.0,NaN,0,Transient,73.80,0,1,Check-Out,2015-07-12
261,Resort Hotel,0,30,2015,July,28,10,2,2,2,...,No Deposit,NaN,NaN,0,Transient-Party,197.00,0,1,Check-Out,2015-07-14
353,Resort Hotel,0,98,2015,July,29,13,1,1,2,...,No Deposit,240.0,NaN,0,Transient-Party,82.00,0,3,Check-Out,2015-07-15
372,Resort Hotel,0,40,2015,July,29,13,1,4,2,...,No Deposit,250.0,NaN,0,Transient,120.00,0,2,Check-Out,2015-07-18


In [9]:
data[data.duplicated()].shape

(31994, 32)

In [10]:
data.duplicated(keep=False).sum()

np.int64(40165)

In [11]:
data.duplicated(keep=False).sum()

np.int64(40165)

In [12]:
print("Number of duplicate rows:", data.duplicated().sum())
data = data.drop_duplicates()

Number of duplicate rows: 31994


In [13]:
print("Shape after removing duplicates:", data.shape)
print("Remaining duplicate rows:", data.duplicated().sum())

Shape after removing duplicates: (87396, 32)
Remaining duplicate rows: 0


### Observations 

-The dataset initially contained 119,390 records, of which 31,994 were identified as exact duplicate rows. These duplicate records were removed to avoid giving disproportionate importance to repeated observations during model training. After removal, the dataset contains 87,396 unique records with no remaining exact duplicates.

In [14]:
print("Missing values in each column:")
print(data.isnull().sum().sort_values(ascending=False))

Missing values in each column:
company                           82137
agent                             12193
country                             452
children                              4
reserved_room_type                    0
assigned_room_type                    0
booking_changes                       0
deposit_type                          0
hotel                                 0
previous_cancellations                0
days_in_waiting_list                  0
customer_type                         0
adr                                   0
required_car_parking_spaces           0
total_of_special_requests             0
reservation_status                    0
previous_bookings_not_canceled        0
is_repeated_guest                     0
is_canceled                           0
distribution_channel                  0
market_segment                        0
meal                                  0
babies                                0
adults                                0
stays_in_

In [15]:
print("Percentage of missing values:")
print((data.isnull().sum() / len(data) * 100).sort_values(ascending=False))

Percentage of missing values:
company                           93.982562
agent                             13.951439
country                            0.517186
children                           0.004577
reserved_room_type                 0.000000
assigned_room_type                 0.000000
booking_changes                    0.000000
deposit_type                       0.000000
hotel                              0.000000
previous_cancellations             0.000000
days_in_waiting_list               0.000000
customer_type                      0.000000
adr                                0.000000
required_car_parking_spaces        0.000000
total_of_special_requests          0.000000
reservation_status                 0.000000
previous_bookings_not_canceled     0.000000
is_repeated_guest                  0.000000
is_canceled                        0.000000
distribution_channel               0.000000
market_segment                     0.000000
meal                               0.000000
ba

In [16]:
data.drop('company', axis=1, inplace=True)

### Observations 

- Since approximately 94% of the values in company are missing, imputing this feature would introduce substantial assumptions and potentially add noise. Therefore, the column is removed from the dataset. 

In [17]:
data['country'].value_counts().head(10)

country
PRT    27453
GBR    10433
FRA     8837
ESP     7252
DEU     5387
ITA     3066
IRL     3016
BEL     2081
BRA     1995
NLD     1911
Name: count, dtype: int64

In [18]:
data[data['children'].isnull()]

,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,booking_changes,deposit_type,agent,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
40600,City Hotel,1,2,2015,August,32,3,1,0,2,...,0,No Deposit,NaN,0,Transient-Party,12.0,0,1,Canceled,2015-08-01
40667,City Hotel,1,1,2015,August,32,5,0,2,2,...,0,No Deposit,14.0,0,Transient-Party,12.0,0,1,Canceled,2015-08-04
40679,City Hotel,1,1,2015,August,32,5,0,2,3,...,0,No Deposit,NaN,0,Transient-Party,18.0,0,2,Canceled,2015-08-04
41160,City Hotel,1,8,2015,August,33,13,2,5,2,...,0,No Deposit,9.0,0,Transient-Party,76.5,0,1,Canceled,2015-08-09


In [19]:
data['country'] = data['country'].fillna('Unknown')

In [20]:
data['country'].isnull().sum()

np.int64(0)

### Observations 

- The country feature contains only 0.52% missing values. Since the missing values cannot be reliably inferred from the available information, they are replaced with an "Unknown" category rather than assigning the most frequent country and introducing potential bias.

In [21]:
data['children'] = data['children'].fillna(0)
data['children'].isnull().sum()

np.int64(0)

### Observations 

- Only 4 observations have missing values in children, representing less than 0.01% of the dataset. Since children is a count-based feature and the number of missing observations is negligible, the missing values are replaced with 0 to retain these records.

In [22]:
missing_values = data.isnull().sum().sort_values(ascending=False)

missing_values[missing_values > 0]

agent    12193
dtype: int64

In [23]:
data['agent'].value_counts().head(15)

agent
9.0      28759
240.0    13028
14.0      3349
7.0       3300
250.0     2779
241.0     1644
28.0      1502
8.0       1383
1.0       1232
6.0       1117
40.0       986
314.0      844
242.0      722
83.0       614
85.0       524
Name: count, dtype: int64

In [24]:
print("Number of unique agents:", data['agent'].nunique())

Number of unique agents: 333


In [25]:
print("Missing agent values:", data['agent'].isnull().sum())

Missing agent values: 12193


In [26]:
data['agent'] = data['agent'].fillna('Unknown').astype(str)

In [27]:
data['agent'].isnull().sum()

np.int64(0)

### Observations 

- The agent feature contains 12,193 missing values (13.95%) and 333 unique travel-agent identifiers. Since agent IDs are categorical identifiers rather than continuous numerical values, numerical imputation such as mean or median would not be meaningful. Missing values are therefore represented as a separate "Unknown" category while retaining the feature.

In [28]:
data.describe().T

,count,mean,std,min,25%,50%,75%,max
is_canceled,87396.0,0.274898,0.446466,0.00,0.0,0.0,1.0,1.0
lead_time,87396.0,79.891368,86.052325,0.00,11.0,49.0,125.0,737.0
arrival_date_year,87396.0,2016.210296,0.686102,2015.00,2016.0,2016.0,2017.0,2017.0
arrival_date_week_number,87396.0,26.838334,13.674572,1.00,16.0,27.0,37.0,53.0
arrival_date_day_of_month,87396.0,15.815541,8.835146,1.00,8.0,16.0,23.0,31.0
stays_in_weekend_nights,87396.0,1.005263,1.031921,0.00,0.0,1.0,2.0,19.0
stays_in_week_nights,87396.0,2.625395,2.053584,0.00,1.0,2.0,4.0,50.0
adults,87396.0,1.875795,0.626500,0.00,2.0,2.0,2.0,55.0
children,87396.0,0.138633,0.455871,0.00,0.0,0.0,0.0,10.0
babies,87396.0,0.010824,0.113597,0.00,0.0,0.0,0.0,10.0


### Observations 

- The statistical summary shows that most numerical features have plausible ranges. However, adults contains zero values and an unusually high maximum of 55, while adr ranges from a negative value of -6.38 to an extreme maximum of 5400. These values require further investigation before deciding whether they represent invalid records or legitimate outliers.

In [29]:
data[data['adults'] == 0].shape

(385, 31)

In [30]:
data[data['adults'] == 0][
    ['hotel', 'adults', 'children', 'babies',
     'stays_in_weekend_nights', 'stays_in_week_nights',
     'adr', 'is_canceled']
].head(20)

,hotel,adults,children,babies,stays_in_weekend_nights,stays_in_week_nights,adr,is_canceled
2224,Resort Hotel,0,0.0,0,0,3,0.0,0
2409,Resort Hotel,0,0.0,0,0,0,0.0,0
3181,Resort Hotel,0,0.0,0,1,2,0.0,0
3684,Resort Hotel,0,0.0,0,1,4,0.0,0
3708,Resort Hotel,0,0.0,0,2,4,0.0,0
4127,Resort Hotel,0,0.0,0,0,0,0.0,1
9376,Resort Hotel,0,0.0,0,0,0,0.0,1
31765,Resort Hotel,0,0.0,0,2,8,28.0,0
32029,Resort Hotel,0,0.0,0,0,1,0.0,0
32827,Resort Hotel,0,0.0,0,3,9,0.0,0


In [31]:
data[(data['adults'] + data['children'] + data['babies']) == 0].shape

(166, 31)

In [32]:
data[(data['adults'] + data['children'] + data['babies']) == 0]['is_canceled'].value_counts()

is_canceled
0    150
1     16
Name: count, dtype: int64

In [33]:
print("Shape after removing invalid guest records:", data.shape)

print(
    "Records with zero guests:",
    ((data['adults'] + data['children'] + data['babies']) == 0).sum()
)

Shape after removing invalid guest records: (87396, 31)
Records with zero guests: 166


### Observations 

- 166 records contained zero adults, children, and babies, meaning no guests were recorded for these bookings. Since such records are not meaningful for analyzing hotel booking behavior, they were treated as invalid observations and removed.

In [34]:
data[data['adr'] < 0][
    ['hotel', 'adults', 'children', 'babies',
     'adr', 'is_canceled', 'reservation_status']
]

,hotel,adults,children,babies,adr,is_canceled,reservation_status
14969,Resort Hotel,2,0.0,0,-6.38,0,Check-Out


In [35]:
data = data[data['adr'] >= 0]

In [36]:
print("Negative ADR values:", (data['adr'] < 0).sum())
print("Shape after removing invalid ADR:", data.shape)

Negative ADR values: 0
Shape after removing invalid ADR: (87395, 31)


### Observations 

- One record contained a negative ADR value of -6.38, which is not logically valid for a hotel room rate. Since this represents an invalid business value rather than a legitimate outlier, the record was removed from the dataset.

In [37]:
data[data['adr'] > 1000][
    ['hotel', 'adults', 'children', 'babies',
     'stays_in_weekend_nights', 'stays_in_week_nights',
     'adr', 'is_canceled']
]

,hotel,adults,children,babies,stays_in_weekend_nights,stays_in_week_nights,adr,is_canceled
48515,City Hotel,2,0.0,0,0,1,5400.0,1


In [38]:
data = data[data['adr'] <= 1000]

### Observations 

- One extreme ADR value of 5400 was identified, while the median ADR was 98.1 and the 75th percentile was 134. Since this single observation was substantially higher than the rest of the distribution and represented a canceled booking, it was treated as an extreme anomaly and removed.

In [39]:
data['total_guests'] = (
    data['adults'] +
    data['children'] +
    data['babies']
)

In [40]:
data[['adults', 'children', 'babies', 'total_guests']].head()

,adults,children,babies,total_guests
0,2,0.0,0,2.0
1,2,0.0,0,2.0
2,1,0.0,0,1.0
3,1,0.0,0,1.0
4,2,0.0,0,2.0


### Observations 

- The adults, children, and babies features describe different guest categories separately. A new total_guests feature was created to represent the overall number of guests in each booking, providing a more direct measure of booking party size

In [41]:
data['total_stay_nights'] = (
    data['stays_in_weekend_nights'] +
    data['stays_in_week_nights']
)

In [42]:
data[
    ['stays_in_weekend_nights',
     'stays_in_week_nights',
     'total_stay_nights']
].head()

,stays_in_weekend_nights,stays_in_week_nights,total_stay_nights
0,0,0,0
1,0,0,0
2,0,1,1
3,0,1,1
4,0,2,2


### Observations 

- The weekend and weekday stay durations were combined to create total_stay_nights, representing the overall length of each booking. The original features were retained because they provide additional information about the distribution of the stay.

In [43]:
data['total_previous_bookings'] = (
    data['previous_cancellations'] +
    data['previous_bookings_not_canceled']
)

In [44]:
data[
    ['previous_cancellations',
     'previous_bookings_not_canceled',
     'total_previous_bookings']
].head()

,previous_cancellations,previous_bookings_not_canceled,total_previous_bookings
0,0,0,0
1,0,0,0
2,0,0,0
3,0,0,0
4,0,0,0


### Observations 

- previous_cancellations and previous_bookings_not_canceled were combined to create total_previous_bookings, representing the customer's overall previous booking history. The original features were retained because they provide distinct information about the customer's past cancellation and successful booking behavior.

In [45]:
data['arrival_date'] = pd.to_datetime(
    data['arrival_date_year'].astype(str) + '-' +
    data['arrival_date_month'] + '-' +
    data['arrival_date_day_of_month'].astype(str),
    format='%Y-%B-%d'
)

In [46]:
data[
    ['arrival_date_year',
     'arrival_date_month',
     'arrival_date_day_of_month',
     'arrival_date']
].head()

,arrival_date_year,arrival_date_month,arrival_date_day_of_month,arrival_date
0,2015,July,1,2015-07-01
1,2015,July,1,2015-07-01
2,2015,July,1,2015-07-01
3,2015,July,1,2015-07-01
4,2015,July,1,2015-07-01


### Observations 

- The year, month, and day components of the arrival date were combined into a single datetime feature, arrival_date. This provides a structured representation of booking arrival dates and enables further extraction of meaningful temporal features.

In [47]:
data['arrival_day_of_week'] = data['arrival_date'].dt.day_name()

In [48]:
data[['arrival_date', 'arrival_day_of_week']].head(10)

,arrival_date,arrival_day_of_week
0,2015-07-01,Wednesday
1,2015-07-01,Wednesday
2,2015-07-01,Wednesday
3,2015-07-01,Wednesday
4,2015-07-01,Wednesday
6,2015-07-01,Wednesday
7,2015-07-01,Wednesday
8,2015-07-01,Wednesday
9,2015-07-01,Wednesday
10,2015-07-01,Wednesday


### Observations 

- The day of the week was extracted from arrival_date to capture potential weekly patterns in hotel bookings. This categorical feature may help the model identify differences in cancellation behavior across different arrival days.

In [49]:
data['is_family'] = (
    (data['children'] > 0) | (data['babies'] > 0)
).astype(int)

### Observations 

- A binary is_family feature was created to identify bookings that include children or babies. This provides a simple indicator of family-oriented bookings that may have different cancellation patterns.

In [50]:
data['has_previous_booking'] = (
    data['total_previous_bookings'] > 0
).astype(int)

### Observations 

- A binary feature was created to indicate whether a customer has any previous booking history, providing a simple measure of customer familiarity with the hotel.

In [51]:
data.drop(
    columns=[
        'reservation_status',
        'reservation_status_date',
        'arrival_date_year',
        'arrival_date_month',
        'arrival_date_day_of_month'
    ],
    inplace=True
)

In [52]:
data.shape

(87394, 33)

In [53]:
data.head()

,hotel,is_canceled,lead_time,arrival_date_week_number,stays_in_weekend_nights,stays_in_week_nights,adults,children,babies,meal,...,adr,required_car_parking_spaces,total_of_special_requests,total_guests,total_stay_nights,total_previous_bookings,arrival_date,arrival_day_of_week,is_family,has_previous_booking
0,Resort Hotel,0,342,27,0,0,2,0.0,0,BB,...,0.0,0,0,2.0,0,0,2015-07-01,Wednesday,0,0
1,Resort Hotel,0,737,27,0,0,2,0.0,0,BB,...,0.0,0,0,2.0,0,0,2015-07-01,Wednesday,0,0
2,Resort Hotel,0,7,27,0,1,1,0.0,0,BB,...,75.0,0,0,1.0,1,0,2015-07-01,Wednesday,0,0
3,Resort Hotel,0,13,27,0,1,1,0.0,0,BB,...,75.0,0,0,1.0,1,0,2015-07-01,Wednesday,0,0
4,Resort Hotel,0,14,27,0,2,2,0.0,0,BB,...,98.0,0,1,2.0,2,0,2015-07-01,Wednesday,0,0


### Observations 

- Outcome-related columns such as reservation_status and reservation_status_date were removed to prevent data leakage. The original arrival date components were also removed after extracting the required temporal features.

### Separate Features and Target

In [54]:
X = data.drop('is_canceled', axis=1)
y = data['is_canceled']

In [55]:
print("Features shape:", X.shape)
print("Target shape:", y.shape)

Features shape: (87394, 32)
Target shape: (87394,)


In [56]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [57]:
print("Training set:", X_train.shape)
print("Testing set:", X_test.shape)

Training set: (69915, 32)
Testing set: (17479, 32)


- The dataset was divided into 80% training and 20% testing data using stratified sampling. Stratification was used to preserve the target-class distribution across both sets.

In [58]:
numerical_cols = X_train.select_dtypes(
    include=['int64', 'float64']
).columns

categorical_cols = X_train.select_dtypes(
    include=['object']
).columns

print("Numerical columns:", list(numerical_cols))
print("\nCategorical columns:", list(categorical_cols))

Numerical columns: ['lead_time', 'arrival_date_week_number', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'booking_changes', 'days_in_waiting_list', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'total_guests', 'total_stay_nights', 'total_previous_bookings', 'is_family', 'has_previous_booking']

Categorical columns: ['hotel', 'meal', 'country', 'market_segment', 'distribution_channel', 'reserved_room_type', 'assigned_room_type', 'deposit_type', 'agent', 'customer_type', 'arrival_day_of_week']


In [59]:
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer

transformer = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), numerical_cols),
        ('cat', OneHotEncoder(handle_unknown='ignore'), categorical_cols)
    ]
)

In [60]:
X_train_transformed = transformer.fit_transform(X_train)
X_test_transformed = transformer.transform(X_test)

In [61]:
print("Transformed training shape:", X_train_transformed.shape)
print("Transformed testing shape:", X_test_transformed.shape)

Transformed training shape: (69915, 574)
Transformed testing shape: (17479, 574)


In [62]:
print("Missing values in X_train:",
      np.isnan(X_train_transformed.toarray()).sum())
      
print("Missing values in X_test:",
      np.isnan(X_test_transformed.toarray()).sum())

Missing values in X_train: 0
Missing values in X_test: 0


In [63]:
print("Training target distribution:")
print(y_train.value_counts(normalize=True))

print("\nTesting target distribution:")
print(y_test.value_counts(normalize=True))

Training target distribution:
is_canceled
0    0.725109
1    0.274891
Name: proportion, dtype: float64

Testing target distribution:
is_canceled
0    0.725099
1    0.274901
Name: proportion, dtype: float64


### Observations 

- The preprocessing pipeline successfully transformed the dataset into 574 model-ready features. Numerical features were standardized and categorical variables were one-hot encoded, while the train-test split was preserved to prevent data leakage.

In [64]:
import os

os.makedirs('../models', exist_ok=True)

In [66]:
import joblib
joblib.dump(transformer, '../models/preprocessor.pkl')

['../models/preprocessor.pkl']

In [67]:
from scipy import sparse
import numpy as np

sparse.save_npz('../models/X_train.npz', X_train_transformed)
sparse.save_npz('../models/X_test.npz', X_test_transformed)

np.save('../models/y_train.npy', y_train)
np.save('../models/y_test.npy', y_test)

In [70]:
import joblib

joblib.dump(transformer, '../models/preprocessor.pkl')

['../models/preprocessor.pkl']

In [71]:
print(transformer.feature_names_in_.tolist())

['hotel', 'lead_time', 'arrival_date_week_number', 'stays_in_weekend_nights', 'stays_in_week_nights', 'adults', 'children', 'babies', 'meal', 'country', 'market_segment', 'distribution_channel', 'is_repeated_guest', 'previous_cancellations', 'previous_bookings_not_canceled', 'reserved_room_type', 'assigned_room_type', 'booking_changes', 'deposit_type', 'agent', 'days_in_waiting_list', 'customer_type', 'adr', 'required_car_parking_spaces', 'total_of_special_requests', 'total_guests', 'total_stay_nights', 'total_previous_bookings', 'arrival_date', 'arrival_day_of_week', 'is_family', 'has_previous_booking']
